src/ThreadedExecutor.h

```diff
@@ -0,0 +1,58 @@
+/* Copyright (C) 2018 European Spallation Source, ERIC. See LICENSE file */
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// \brief Simple threaded executor to simplify the use of threading in this lib
+/// library.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <concurrentqueue/concurrentqueue.h>
+#include <functional>
+#include <future>
+#include <memory>
+#include <thread>
+
+class ThreadedExecutor {
+private:
+public:
+  using WorkMessage = std::function<void()>;
+  ThreadedExecutor(bool LowPrioThreadExit = false)
+      : LowPrioExit(LowPrioThreadExit), WorkerThread(ThreadFunction) {}
+  ~ThreadedExecutor() {
+    if (LowPrioExit) {
+      SendLowPrioWork([=]() { RunThread = false; });
+    } else {
+      SendWork([=]() { RunThread = false; });
+    }
+    WorkerThread.join();
+  }
+  void SendWork(WorkMessage Message) { MessageQueue.enqueue(Message); }
+  size_t size_approx() { return MessageQueue.size_approx(); }
+  void SendLowPrioWork(WorkMessage Message) {
+    LowPrioMessageQueue.enqueue(Message);
+  }
+
+private:
+  bool RunThread{true};
+  std::function<void()> ThreadFunction{[=]() {
+    while (RunThread) {
+      WorkMessage CurrentMessage;
+      if (MessageQueue.try_dequeue(CurrentMessage)) {
+        CurrentMessage();
+      } else if (LowPrioMessageQueue.try_dequeue(CurrentMessage)) {
```

Comment:
What's the idea for the priority levels here, I don't think we discussed it?
It isn't used in tests etc makes me particularly nervous.

Reply:
To be fair, it is a bit of a bodge. The reason is that `moodycamel::ConcurrentQueue` does not give **ANY** guarantees about ordering of items produced from multiple threads. Thus, to prevent the thread from exiting before everything is written to file, a low priority message queue is used.

This is one reason to forgo the use of `ThreadedExecutor`.

Reply:
By the way: Seeing as I know that you rarely read the PR-description: please comment on the use of `ThreadedExecutor`.

Reply:
I'm happy to proceed with this design. A little concerned with possible overhead of `std::function` for every message, but I suggest we do some performance review, profiling etc of the v3 branch before it hits master.

Reply:
Ok, thanks!